File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        def dfs(node):
            if not node:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)
            if(abs(left-right)>1):
                raise ValueError
            return 1 + max(left,right)
        try:
            dfs(root)
            return True
        except ValueError:
            return False

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height(TreeNode* root){
        if(!root)
            return 0;
        return 1+max(height(root->right),height(root->left));
    }
    bool isBalanced(TreeNode* root) {
        if(!root)
            return true;
        int leftHeight = height(root->left);
        int rightHeight = height(root->right);
        
        if(abs(leftHeight-rightHeight)>1)
            return false;
        
        if(!isBalanced(root->left) || !isBalanced(root->right))
            return false;
        return true;
    }
};